### PR TITLE
fix: shared user session crashes WPB-4920

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-4920" title="WPB-4920" target="_blank">BugWPB-4920</a>  [iOS] Crash when sending message in C1
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This is a feature branch for a collection of refactorings needed to prevent crashes when trying to access the shared user session.

### Issues

Often the app crashes shortly after logging out or the session expiring.

### Causes

The UI code depends heavily on the active user session, which is accessed via a singleton `ZMUserSession.shared()`. This is often safely unwrapped, but there are places where it is not, leading to a a crash. The most common point of crash is when the the static function `ZMUser.selfUser()` is accessed, which depends on the shared user session and implicitly unwraps its value.

In theory, the UI code should be running when there is no user session, but this can not be guaranteed, because the UI code may not have been torn down yet, or part of it may be retained and fired as a block in response to a notification or other observation mechanism.

### Solutions

Avoid all static access to the shared user session and self user. Instead, inject a non optional value from the top level and propagate it where it needs to go via dependency injection.

### Note

Any refinement to better model the user session and dependency structure isn't considered here. First we want to remove the static accessors and make it work. After we can polish it.

#### Test Coverage

- [ ] Updated tests.

#### How to Test

- Be logged in.
- Remove your current device from another device, causing the current device to be logged out
- Assert you don't crash

You may need to repeat this a few times.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Source/Model/FeatureConfig/AppLock/AppLockController.swift [hunk 1]
```diff
@@ -116,7 +116,11 @@ public final class AppLockController: AppLockType {
 
     // MARK: - Life cycle
 
-    public init(userId: UUID, selfUser: ZMUser, legacyConfig: LegacyConfig? = nil) {
+    public init(
+        userId: UUID,
+        selfUser: ZMUser,
+        legacyConfig: LegacyConfig?
```
Comment (caldrian) [inline wire-ios-data-model/Source/Model/FeatureConfig/AppLock/AppLockController.swift:122]: I like that you removed the default value, I think default values can lead to unsafe code.

**Review by netbe — APPROVED** [review]

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+
+    func enqueue(
+        _ changes: @escaping () -> Void,
+        completionHandler: (() -> Void)?
+    )
+
+    // TODO: rename to "shouldHideNotificationContent"
+    var isNotificationContentHidden: Bool { get set }
+
+    // TODO: rename to "isEncryptionAtRestEnabled"
```
Comment (caldrian) [inline wire-ios-sync-engine/Source/UserSession/UserSession.swift:135]: Should we add a ticket, a date or some other reference for when these `TODOs` should be solved?

**Review by johnxnguyen — COMMENTED** [review]
@agisilaos do you still need to revert the fix for Xcode 15 errors?

**Comment by agisilaos** [comment]
> @agisilaos do you still need to revert the fix for Xcode 15 errors?

@johnxnguyen Yep! Those changes have been reverted.

File: wire-ios-sync-engine/Source/UserSession/UserSession.swift (hunk 2 above)
Comment (agisilaos) [reply wire-ios-sync-engine/Source/UserSession/UserSession.swift:135]: I will take care of it today, while I create the tickets for the rest of the work related to the userSession.

File: wire-ios/Wire-iOS Tests/ConversationMessageCell/ConversationMessageSnapshotTestCase.swift [hunk 3]
```diff
@@ -32,7 +33,26 @@ extension ConversationMessageContext {
                                                                        spacing: 0)
 }
 
-extension XCTestCase {
+func XCTAssertArrayEqual(_ descriptions: [Any], _ expectedDescriptions: [Any], file: StaticString = #file, line: UInt = #line) {
```
Comment (caldrian) [inline wire-ios/Wire-iOS Tests/ConversationMessageCell/ConversationMessageSnapshotTestCase.swift:36]: I suggest to remove this helper. I think it is already being used in a wrong way. I will create a PR, if you agree? @agisilaos

Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/ConversationMessageCell/ConversationMessageSnapshotTestCase.swift:36]: I can take a look on how it's being used but sure, please do open a separate PR. If not please do create a ticket so we can tackle it later.

**Review by agisilaos — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/Components/Settings/ZMUserSessionInterface.swift [hunk 4]
```diff
@@ -20,6 +20,7 @@ import WireSyncEngine
 
 typealias UserSessionInterface = UserSessionSwiftInterface & UserSessionAppLockInterface
 
+// TODO: delete
```
Comment (caldrian) [inline wire-ios/Wire-iOS/Sources/Components/Settings/ZMUserSessionInterface.swift:23]: Same here, could we specify when and/or why this needs to be done?

**Pull request merged**